### PR TITLE
Hc networking option

### PR DESCRIPTION
**Pull request by sphinxc0re**

### CODE REVIEW

**Comment by zippy** [comment]
Ok, I got this working in the very super simplest of ways.  Namely you can activate networking by simply supplying an environment variable with the path to your n3h instance.  Eg from my terminal:

``` shell
HC_N3H_PATH=/home/eric/code/metacurrency/holochain/n3h hc run
Spawning network with working directory: /tmp
SPAWN ("node" "/home/eric/code/metacurrency/holochain/n3h/packages/n3h/bin/n3h")
(@hackmode@) [t] bound to tcp://127.0.0.1:45875
(@hackmode@) [i] p2p bound [
  "/ip4/127.0.0.1/tcp/44753/ipfs/QmdWBR5jEMJRtCCzqGS6vgXjaRknAcsTU1VaYoCiqMw4ES",
  "/ip4/192.168.1.5/tcp/44753/ipfs/QmdWBR5jEMJRtCCzqGS6vgXjaRknAcsTU1VaYoCiqMw4ES"
]
(@hackmode@) [t] running
READY! tcp://127.0.0.1:45875 ["/ip4/127.0.0.1/tcp/44753/ipfs/QmdWBR5jEMJRtCCzqGS6vgXjaRknAcsTU1VaYoCiqMw4ES", "/ip4/192.168.1.5/tcp/44753/ipfs/QmdWBR5jEMJRtCCzqGS6vgXjaRknAcsTU1VaYoCiqMw4ES"]
Network spawned with binding: "tcp://127.0.0.1:45875"
2019-01-08 22:13:33:test-instance: debug/reduce: ActionWrapper { action: InitApplication(Dna.

.. snip..
```

File: cmd/src/cli/run.rs [hunk 1]
```diff
@@ -65,11 +65,30 @@ pub fn run(package: bool, port: u16, persist: bool) -> DefaultResult<()> {
         rules,
     };
 
+    let n3h_path = env::var("HC_N3H_PATH").ok();
+    let n3h_mode = env::var("HC_N3H_MODE").ok();
+    let n3h_persistence_path = env::var("HC_N3H_WORK_DIR").ok();
+
+    // network config
+    let network_config = if networked || n3h_path.is_some() {
```
Comment (sphinxc0re) [inline cmd/src/cli/run.rs]: Why is it relevant, wether the n3h path is set when we will take the default anyway?

File: cmd/README.md [hunk 2]
```diff
@@ -215,6 +220,37 @@ This will store data in the same directory as your app, in a hidden folder calle
...
+  "/ip4/192.168.1.5/tcp/34199/ipfs/QmTg9qMFBosfWD8yeLbcNUwT8UgwNKoT9mGEfm9vXKEHzS"
+]
+(@hackmode@) [t] running
+...
+```
+
+To fire up a second node you have to do a little more work, namely:
+1. providing the address of the first node as a bootstrap node,
+2. specifying a different agent id
+3. specifying a different port for the UI interface.
```
Comment (Connoropolous) [inline cmd/README.md]:
```suggestion
3. specifying a different port for the websocket server, for a UI to connect to.
```

File: cmd/README.md [hunk 3]
```diff
@@ -215,6 +220,37 @@ This will store data in the same directory as your app, in a hidden folder calle
...
+SPAWN ("node" "/home/eric/holochain/n3h/packages/n3h/bin/n3h")
+(@hackmode@) [t] bound to tcp://127.0.0.1:42341
+(@hackmode@) [i] p2p bound [
+  "/ip4/127.0.0.1/tcp/34199/ipfs/QmTg9qMFBosfWD8yeLbcNUwT8UgwNKoT9mGEfm9vXKEHzS",
+  "/ip4/192.168.1.5/tcp/34199/ipfs/QmTg9qMFBosfWD8yeLbcNUwT8UgwNKoT9mGEfm9vXKEHzS"
+]
+(@hackmode@) [t] running
+...
+```
+
```
Comment (Connoropolous) [inline cmd/README.md]:
```suggestion
Note that there is an agent id set by default, and the default is `testAgent`.
```

File: container_api/src/config.rs [hunk 4]
```diff
@@ -335,29 +333,37 @@ pub struct Bridge {
...
     pub n3h_ipc_uri: Option<String>,
 }
 
-fn default_n3h_mode() -> String {
+pub fn default_n3h_mode() -> String {
     String::from("HACK")
 }
 
+pub fn default_n3h_path() -> String {
+    String::from("~/.hc/net/n3h")
```
Comment (Connoropolous) [inline container_api/src/config.rs]: @zippy I think this is another one of those paths that wouldn't work on windows, isn't it?

File: cmd/README.md [hunk 5]
```diff
@@ -215,6 +220,37 @@ This will store data in the same directory as your app, in a hidden folder calle
 
 Of course these options can be used in combination with one another.
 
+### Networking
+
+`hc run` uses mock networking by default and therefore can't talk to any other nodes.  If you want to test multiple nodes you will need to install the [n3h](https://github.com/holochain/n3h) networking component (following the instructions on the readme there).  Once you have installed it then you can simply fire up your first node while setting the HC_N3H_PATH environment variable to the path where you installed it.  If n3h was installed properly you should see something like this:
+
+``` shell
+$ HC_N3H_PATH=/home/eric/holochain/n3h hc run
```
Comment (Connoropolous) [inline cmd/README.md]: Don't you have to add `--networked` ? is the default for it be enabled or not? 

I can't see in the code a default for the `networked` variable

Comment (zippy) [reply cmd/README.md]: you don't I just made it work with the presence of the environment variable.  I didn't even add the --networked flag for now.

Comment (Connoropolous) [reply cmd/README.md]: @zippy this is the code for a `--networked` option
https://github.com/holochain/holochain-rust/pull/826/files#diff-3f97b1cf8c268f48439642916602b667R113

File: container_api/src/config.rs (hunk 4 above)
Comment (zippy) [reply container_api/src/config.rs]: working on a fix for this, don't know what the right directory is in the windows case...

File: cmd/README.md (hunk 5 above)
Comment (zippy) [reply cmd/README.md]: well, would you look at that!  I guess Julian did that before I started working on it.  You still have to set the ENV vars for now though... hopefully in the future that will go away.

File: container_api/src/config.rs (hunk 4 above)
Comment (zippy) [reply container_api/src/config.rs]: fixed and pushed

File: cmd/src/main.rs [hunk 6]
```diff
@@ -109,6 +109,8 @@ enum Cli {
         package: bool,
         #[structopt(long, help = "Save generated data to file system")]
         persist: bool,
+        #[structopt(long, help = "Use real networking")]
```
Comment (Connoropolous) [inline cmd/src/main.rs]: Since this will show up in the help text, should probably improve this very unhelpful explanation... thinking...

**Pull request merged**